From 1bfac318da67c2a0c0325f30a16cdd23ac7408a9 Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Sat, 5 May 2018 12:51:42 +0200
Subject: [PATCH 1/3] Allow providing flex and bison binaries

* The prebuilt flex binary crashes due changes in newer glibc versions,
  and the prebuilt bison binary doesn't work on newer Darwin versions.
  -> Add a flag to allow the user to provide its own flex and bison binary,
     that should match the version provided by AOSP.

Usage: Add `export FLEX_EXEC=$path_to_flex` in your .bashrc/.zshrc
Example: `export FLEX_EXEC=$HOME/android/flex-2.5.39`

Change-Id: Ia93572e54accee1bb07f47f77ec8e159695b0e25

# Conflicts:
#	cc/gen.go
---
 cc/config/global.go | 14 ++++++++++++++
 cc/gen.go           |  4 ++--
 2 files changed, 16 insertions(+), 2 deletions(-)

diff --git a/cc/config/global.go b/cc/config/global.go
index 7a2aa80f..0e07badf 100644
--- a/cc/config/global.go
+++ b/cc/config/global.go
@@ -228,6 +228,20 @@ func init() {
 		}
 		return ""
 	})
+
+	pctx.VariableFunc("FlexExec", func(ctx android.PackageVarContext) string {
+		if override := ctx.Config().Getenv("FLEX_EXEC"); override != "" {
+			return override
+		}
+		return "prebuilts/misc/${HostPrebuiltTag}/flex/flex-2.5.39"
+	})
+
+	pctx.VariableFunc("BisonExec", func(ctx android.PackageVarContext) string {
+		if override := ctx.Config().Getenv("BISON_EXEC"); override != "" {
+			return override
+		}
+		return "prebuilts/build-tools/${HostPrebuiltTag}/bin/bison"
+	})
 }
 
 var HostPrebuiltTag = pctx.VariableConfigMethod("HostPrebuiltTag", android.Config.PrebuiltOS)
diff --git a/cc/gen.go b/cc/gen.go
index c794f5c3..0c8ae074 100644
--- a/cc/gen.go
+++ b/cc/gen.go
@@ -25,8 +25,8 @@ import (
 )
 
 func init() {
-	pctx.SourcePathVariable("lexCmd", "prebuilts/build-tools/${config.HostPrebuiltTag}/bin/flex")
-	pctx.SourcePathVariable("yaccCmd", "prebuilts/build-tools/${config.HostPrebuiltTag}/bin/bison")
+	pctx.SourcePathVariable("lexCmd", "${config.FlexExec}")
+	pctx.SourcePathVariable("yaccCmd", "${config.BisonExec}")
 	pctx.SourcePathVariable("yaccDataDir", "prebuilts/build-tools/common/bison")
 
 	pctx.HostBinToolVariable("aidlCmd", "aidl-cpp")
-- 
2.17.1

